src/print.rs

```diff
@@ -0,0 +1,318 @@
+                        block_name_sep,
+                        index,
+                        comment_end,
+                    )
+                    .unwrap();
+                }
+                result.push_str(&block.compile(&code_blocks, settings)?);
+                result.push('\n');
+                if !clean && (idx == blocks.len() - 1 || block.name != blocks[idx + 1].name) {
+                    write!(
+                        result,
+                        "{} {}{}{}{}{}{}{}",
+                        comment_start,
+                        block_end,
+                        path,
+                        block_name_sep,
+                        name,
+                        block_name_sep,
+                        index,
+                        comment_end,
+                    )
+                    .unwrap();
+                }
+            }
+        }
+        None => {
+            return Err(CompileError::Single {
+                line_number: 0,
+                kind: CompileErrorKind::MissingEntrypoint,
+            })
+        }
+    }
+    if settings.map(|s| s.eof_newline).unwrap_or(true) && !result.ends_with('\n') {
+        result.push('\n');
+    }
+    Ok(result)
+}
+
+/// Formats this `Document` as the original source, potentially replacing code blocks
+pub fn print_reverse(
+    document: &Document,
+    settings: &ParserSettings,
+    code_blocks: &HashMap<(&Option<String>, &usize), &RevCodeBlock>,
+) -> String {
+    let mut block_count: HashMap<&Option<String>, usize> = HashMap::new();
+
+    let mut output = String::new();
+    for node in &document.nodes {
+        match node {
+            Node::Transclusion(transclusion) => {
+                output.push_str(&print_transclusion(transclusion, true))
+            }
+            Node::Text(text_block) => {
+                output.push_str(&text_block.to_string());
+                output.push('\n');
+            }
+            Node::Code(code_block) => {
+                let index = {
+                    let count = block_count.entry(&code_block.name).or_default();
+                    *count += 1;
+                    *count - 1
+                };
+
+                let alt_block = code_blocks.get(&(&code_block.name, &index));
+
+                output.push_str(
+                    &print_code_block_reverse(code_block, alt_block.copied(), settings)
+                        .split('\n')
+                        .map(|line| {
+                            if line.is_empty() {
+                                line.to_string()
+                            } else {
+                                format!("{}{}", code_block.indent, line)
+                            }
+                        })
+                        .collect::<Vec<_>>()
+                        .join("\n"),
+                )
+            }
+        }
+    }
+    output
+}
+
+pub fn print_transclusion(transclusion: &Transclusion, reverse: bool) -> String {
```

Comment:
This looks suspiciously like two unrelated functions shoved into one using a `bool` flag.